### DISCUSSION TITLE
Please recommend a front-end development framework that works best with FastAPI and is based on Python.

### DISCUSSION THREAD

**Discussion by uuleejie** [fastapi/fastapi#11644] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app=FastAPI()

@app.get("/")
def root():
    return {"message":"Hello world"}
```

### Description

I really like Fastapi and plan to use it to develop my website. Since I can only write python code, most Javascript-based front-end development frameworks give me a headache. I hope that the FastAPI experts will recommend me a Python-based front-end development framework that works well with FastAPI, and that's all right! Thank you!

### Operating System

Windows

### FastAPI Version

0.101.1

### Pydantic Version

2.6.1

### Python Version

3.12.3

**Comment by visuxls** [comment]
You could give [FastUI](https://github.com/pydantic/FastUI) a try.

**Comment by n8sty** [comment]
HTMX

**Comment by hasansezertasan** [comment]
TL;DR: If you just want to built a simple website with so few pages, just go with HTMX and Tailwind.

You could go with one of the two: NiceGUI and Reflex. They are built on top of FastAPI, so you will be able to write python code and get a web application based on FastAPI.

You could go with HTMX and TailwindCSS but they work better with Django, since FastAPI is more focused on the API development side of things. If you choice to go with HTMX and TailwindCSS but don't like django, I could recommend one of three: Flask (has a lot of awesome extensions), Litestar (has built in HTMX support), and Starlette (lighter then two other).

There is also another tool called FastUI (I have contributed to it lately). It's a good tool for simple, (very simple) tasks.

One final recommendation could be using something like Flask Admin (you can directly mount a flask app to fastapi app). What do I mean by that? Flask Admin comes with automatic menu highlighting, ready page layouts, Class Based View support, support for various ORMs and ODM alongside with a file admin and simple Redis console. These features are the tip of the mountain.

I have a project called [not-just-a-todo-app](https://www.github.com/hasansezertasan/not-just-a-todo-app). It's built with Flask Admin. It's a POC.

If you like the idea of building an application through an Admin Interface (I have deployed some production applications with this method) but you still want to stay in FastAPI, you have many options: SQLAdmin, FastAPI-Admin, FastAPI-Amis-	Admin, and Starlette Admin, etc.

In general, it depends on your needs. I happen to know this many tools and approaches by years of experience and hunger for more python.

**Reply by uuleejie** [reply]
Thank you for your detailed answers. I think I'd consider using FastUI as a front-end development framework to work with FastAPI at the moment, as they have a lot in common. For example, they are all based on pydantic. But FastUI is really too young, and there isn't enough documentation and community support to use it. I hope that FastUI will become more and more powerful in the future.

**Reply by hasansezertasan** [reply]
> Thank you for your detailed answers.

You're welcome, happy to help 🤗

> I think I'd consider using FastUI as a front-end development framework to work with FastAPI at the moment, as they have a lot in common. For example, they are all based on pydantic.

You're right, but don't forget, because it's quite young and it's built on top of react, you might not be able to build the interface you want with it.

> But FastUI is really too young, and there isn't enough documentation and community support to use it.

I have another repository called [fastui-tutorials](https://www.github.com/hasansezertasan/fastui-tutorials) which contains some entry level examples. You can check that out to start with.

Lack of documentation is a real minus 😞.

> I hope that FastUI will become more and more powerful in the future.

Taught and prayers goea to all python technologies.

**Reply by uuleejie** [reply]
> I have another repository called [fastui-tutorials](https://www.github.com/hasansezertasan/fastui-tutorials) which contains some entry level examples. You can check that out to start with.

That's great.

**Comment by mmmcorpsvit** [comment]
reflex?

**Comment by uuleejie** [comment]
Maybe. I am studying it.

________________________________
发件人: Nikolas ***@***.***>
发送时间: 2024年12月9日 17:35
收件人: fastapi/fastapi ***@***.***>
抄送: Leo Lee ***@***.***>; Author ***@***.***>
主题: Re: [fastapi/fastapi] Please recommend a front-end development framework that works best with FastAPI and is based on Python. (Discussion #11644)


reflex?

―
